src/wp-includes/media.php

```diff
@@ -5564,6 +5564,121 @@ function wp_get_loading_attr_default( $context ) {
+	if ( 'template' === $context ) {
+		// TO CHECK:
+		return $loading_attrs;
+	}
+
+	// If the count has already reached or exceeded the threshold, all of the images after can be lazy-loaded.
+	$content_media_count = wp_increase_content_media_count( 0 );
+	if ( wp_omit_loading_attr_threshold() < $content_media_count ) {
+		$loading_attrs['loading'] = 'lazy';
+		return $loading_attrs;
+	}
+
+	$img_size = $attr['width'] * $attr['height'];
+
+	// Do not lazy-load images in the header block template part, as they are likely above the fold.
+	// For classic themes, this is handled in the condition below using the 'get_header' action.
+	$header_area = WP_TEMPLATE_PART_AREA_HEADER;
+	if ( "template_part_{$header_area}" === $context ) {
+		if ( WP_LCP_MIN_IMAGE_SIZE <= $img_size && 1 === wp_increase_content_media_count() ) {
+			$loading_attrs['fetchpriority'] = 'high';
+		}
+		return $loading_attrs;
+	}
+
+	// Special handling for programmatically created image tags.
+	if ( ( 'the_post_thumbnail' === $context || 'wp_get_attachment_image' === $context ) ) {
+		/*
+		 * Skip programmatically created images within post content as they need to be handled together with the other
+		 * images within the post content.
+		 * Without this clause, they would already be counted below which skews the number and can result in the first
+		 * post content image being lazy-loaded only because there are images elsewhere in the post content.
+		 */
+		if ( doing_filter( 'the_content' ) ) {
+			if ( WP_LCP_MIN_IMAGE_SIZE <= $img_size && 1 === wp_increase_content_media_count() ) {
+				$loading_attrs['fetchpriority'] = 'high';
+			}
+			return $loading_attrs;
+		}
+
+		// Conditionally skip lazy-loading on images before the loop.
+		if (
+			// Only apply for main query but before the loop.
+			$wp_query->before_loop && $wp_query->is_main_query()
+			/*
+			 * Any image before the loop, but after the header has started should not be lazy-loaded,
+			 * except when the footer has already started which can happen when the current template
+			 * does not include any loop.
+			 */
+			&& did_action( 'get_header' ) && ! did_action( 'get_footer' )
+		) {
+			if ( WP_LCP_MIN_IMAGE_SIZE <= $img_size && 1 === wp_increase_content_media_count() ) {
+				$loading_attrs['fetchpriority'] = 'high';
+			}
+			return $loading_attrs;
+		}
+	}
+
+	/*
+	 * The first elements in 'the_content' or 'the_post_thumbnail' should not be lazy-loaded,
+	 * as they are likely above the fold.
+	 */
+	if ( 'the_content' === $context || 'the_post_thumbnail' === $context ) {
+		// Only elements within the main query loop have special handling.
+		if ( is_admin() || ! in_the_loop() || ! is_main_query() ) {
+			$loading_attrs['loading'] = 'lazy';
+			return $loading_attrs;
+		}
+
+		if ( WP_LCP_MIN_IMAGE_SIZE <= $img_size ) {
+			// Increase the counter since this is a main query content element.
+			$content_media_count = wp_increase_content_media_count();
```

Comment:
Need to revisit the logic here. Since we are using `wp_increase_content_media_count` for both lazy loading and fetchpriority we will need to check image size before incrementing for both lazy-loading and fetchpriority.